src/replica/Configuration.cc

```diff
@@ -341,13 +376,47 @@ void Configuration::_load(util::Lock const& lock, string const& configUrl, bool
     _data["database"]["password"] = _connectionParams.password;
     _data["database"]["name"] = _connectionParams.database;
 
+    // The schema upgrade timer is used for limiting a duration of time when
+    // tracking (if enabled) the schema upgrade. The timeout includes
+    // the connect (or reconnect) time.
+    util::Timer schemaUpgradeTimer;
+    schemaUpgradeTimer.start();
+
     // Read data, validate and update configuration parameters.
     _connectionPtr = database::mysql::Connection::open(_connectionParams);
-    _connectionPtr->executeInOwnTransaction([&](decltype(_connectionPtr) conn) {
-        ConfigParserMySQL parser(conn, _data, _workers, _databaseFamilies, _databases);
-        parser.parse();
-    });
-
+    while (true) {
+        try {
+            _connectionPtr->executeInOwnTransaction([&](decltype(_connectionPtr) conn) {
+                ConfigParserMySQL parser(conn, _data, _workers, _databaseFamilies, _databases);
+                parser.parse();
+            });
+            break;
+        } catch (ConfigVersionMismatch const& ex) {
+            if (Configuration::schemaUpgradeWait()) {
+                if (ex.version > ex.requiredVersion) {
+                    LOGS(_log, LOG_LVL_ERROR, _context() << "Database schema version is newer than"
+                        << " the one required by the application, ex: " << ex.what());
+                        throw;
+                }
+                schemaUpgradeTimer.stop();
```

Comment:
do you need to start the timer again?

Reply:
Not in this case. The name of the class `util::Time` that's I'm using is confusing. In reality, it's not a "timer" (the method won't start any system timer, like asynchronous timers in BOOST ASIO). This is just a simple utility class for capturing and measuring time intervals by calling methods:
```
start() - capture the current time and store it in a member variable
stop() - capture the current time and store it in another member variable
```
And there is another method that computes a difference (an interval passed) between those two member variables:
```
timeDiff()
```
That's why I'm calling `start()` just once at the very beginning of waiting for the schema upgrade tracking loop. Then I'm calling `stop()` after each failed check in order to "extend" the interval of time stored in that "Timer" object. And this is going on until the "timer" "expires".
Norte details can be found in the header of the class:
https://github.com/lsst/qserv/blob/main/src/util/Timer.h